src/reducers/notifications.ts

```diff
@@ -0,0 +1,52 @@
+import { NotifyClientTypes } from '@walletconnect/notify-client'
+
+export interface TopicNotificationsState {
+  fullNotifications: NotifyClientTypes.NotifyMessage[]
+  existingIds: Set<string>
+  hasMore: boolean
+}
+
+export interface NotificationsState {
+  [topic: string]: TopicNotificationsState
+}
+
+export type NotificationsActions = {
+  type: 'FETCH_NOTIFICATIONS'
+  notifications: NotifyClientTypes.NotifyMessage[]
+  topic: string
+  hasMore: boolean
+}
+
+// Opted for a reducer since the state changes are complex enough to warrant
+// changes to a set and an array and having all that inside the hooks would
+// cause too many updates to the hooks, causing unnecessary rerenders.
+export const notificationsReducer = (
+  state: NotificationsState,
+  action: NotificationsActions
+): NotificationsState => {
+  const topicState = state[action.topic] as TopicNotificationsState | undefined
+
+  switch (action.type) {
+    case 'FETCH_NOTIFICATIONS':
+      const ids = topicState?.existingIds || new Set<string>()
+      const filteredVals = action.notifications.filter(val => !ids.has(val.id))
+      const uniqueIds = action.notifications.map(notification => notification.id)
+
+      const fullNotifications = topicState?.fullNotifications || []
+      const newFullIdsSet = new Set(topicState?.existingIds || [])
+
+      for (const val of uniqueIds) {
+        newFullIdsSet.add(val)
```

Comment:
I think you should be able to simplify this whole section into one `Set` assignment unless I'm missing a step:

```
const newFullIdsSet = new Set([...array1, ...array2]);
```

Reply:
This would be a bit too slow for my liking, if hypothetically scrolling down a list of a lot of GMs or something. We get minor optimization for no noticable reduction in readability